### PR TITLE
Feature: A better share button NeedMoreShare2

### PR DESCRIPTION
**Pull request by LEAFERx**
## PR Checklist
**Please check if your PR fulfills the following requirements:**

- [x] The commit message follows [our guidelines](https://github.com/iissnan/hexo-theme-next/blob/master/.github/CONTRIBUTING.md).
- [x] Tests for the changes have been added (for bug fixes / features).
   - [x] Muse | Mist have been tested.
   - [x] Pisces | Gemini have been tested.
- [ ] Docs have been added / updated (for bug fixes / features).

## PR Type
**What kind of change does this PR introduce?**  

- [ ] Bugfix.
- [x] Feature.
- [ ] Code style update (formatting, local variables).
- [ ] Refactoring (no functional changes, no api changes).
- [ ] Build related changes.
- [ ] CI related changes.
- [ ] Documentation content changes.
- [ ] Other... Please describe:

## What is the new behavior?
Add a new share button which is a pure javascript lib.
It will be the only sharebutton that supports HTTPS in NexT (for now) and it has a better UI.
It's useful in China.
https://github.com/revir/need-more-share2

* Screens with this changes: 
### 1 Desktop Postbottom
image 1
when hover: 
image 2
when click: 
image 3
### 2 Desktop Float
image 4
when hover:
image 5
when click: 
image 6
### 3 Mobile Postbottom
image 11
image 12
### 4 Mobile Float
image 9
image 10

### How to use?
In NexT `_config.yml`:
```yml
# NeedMoreShare2
# This plugin is a pure javascript sharing lib which is useful in China.
# See: https://github.com/revir/need-more-share2
needMoreShare2:
  postbottom: 
    enable: true
    Options:
      iconStyle: box
      boxForm: horizontal
      position: middleCenter
      networks: Weibo,Wechat,Douban,QQZone,Twitter,Facebook
  float: 
    enable: true
    Options:
      iconStyle: box
      boxForm: horizontal
      position: middleRight
      networks: Weibo,Wechat,Douban,QQZone,Twitter,Facebook
```

## Does this PR introduce a breaking change?
- [ ] Yes.
- [x] No.

### CODE REVIEW

File: source/css/_common/components/third-party/third-party.styl [hunk 1]
```diff
@@ -5,3 +5,4 @@
 @import "localsearch";
 @import "busuanzi-counter";
 @import "algolia-search" if hexo-config('algolia_search.enable');
+@import "needsharebutton";
```
Comment (ivan-nginx) [inline source/css/_common/components/third-party/third-party.styl]: Need to add conditional import here.

File: _config.yml [hunk 2]
```diff
@@ -390,6 +390,25 @@ valine:
 # Share
 #duoshuo_share: true
 
+# NeedMoreShare2
+# This plugin is a pure javascript sharing lib which is useful in China.
+# See: https://github.com/revir/need-more-share2
+needMoreShare2:
+  postbottom:
```
Comment (ivan-nginx) [inline _config.yml]: Need to add global enable/disable switch for better control it.

**Comment by ivan-nginx** [comment]
@LEAFERx and can u show live demo on it? Why on your blog i don't see this feature?

**Comment by LEAFERx** [comment]
@ivan-nginx now it's on my blog

File: _config.yml [hunk 3]
```diff
@@ -390,6 +390,26 @@ valine:
...
 #duoshuo_share: true
 
+# NeedMoreShare2
+# This plugin is a pure javascript sharing lib which is useful in China.
+# See: https://github.com/revir/need-more-share2
+needMoreShare2:
+  enable: false
+  postbottom: 
+    enable: false
+    Options:
```
Comment (ivan-nginx) [inline _config.yml]: To lowercase?

File: _config.yml [hunk 4]
```diff
@@ -390,6 +390,26 @@ valine:
...
+      iconStyle: box
+      boxForm: horizontal
+      position: middleCenter
+      networks: Weibo,Wechat,Douban,QQZone,Twitter,Facebook
+  float: 
+    enable: false
+    Options:
```
Comment (ivan-nginx) [inline _config.yml]: To lowercase?

File: _config.yml [hunk 5]
```diff
@@ -390,6 +390,26 @@ valine:
 # Share
 #duoshuo_share: true
 
+# NeedMoreShare2
+# This plugin is a pure javascript sharing lib which is useful in China.
+# See: https://github.com/revir/need-more-share2
+needMoreShare2:
```
Comment (ivan-nginx) [inline _config.yml]: To lowercase? (_)

**Comment by ivan-nginx** [comment]
Also, need to test it on other schemes (Pisces | Gemini).

File: layout/_third-party/share/needsharebutton/floatdom.swig [hunk 6]
```diff
@@ -0,0 +1 @@
+<div id="need-share-button-float"><i class="fa fa-share-alt" aria-hidden="true"></i></div>
```
Comment (ivan-nginx) [inline layout/_third-party/share/needsharebutton/floatdom.swig]: And can u please to split this 2 files into 1?

File: layout/_third-party/share/needsharebutton/postbottomdom.swig [hunk 7]
```diff
@@ -0,0 +1 @@
+<span id="need-share-button-postbottom"><i class="fa fa-share-alt" aria-hidden="true"></i></span>
```
Comment (ivan-nginx) [inline layout/_third-party/share/needsharebutton/postbottomdom.swig]: And can u please to split this 2 files into 1?

**Comment by ivan-nginx** [comment]
Also, need to add all supported networks from [needShareButton](https://github.com/DzmVasileusky/needShareButton):
```
default: 'Mailto,Twitter,Pinterest,Facebook,GooglePlus,Reddit,Delicious,Tapiture,StumbleUpon,Linkedin,Slashdot,Technorati,Posterous,Tumblr,GoogleBookmarks,Newsvine,Pingfm,Evernote,Friendfeed,Vkontakte,Odnoklassniki,Mailru'
options: Any of the network name's from pervios row comma separated
```

File: _config.yml [hunk 8]
```diff
@@ -390,6 +390,26 @@ valine:
...
+  enable: false
+  postbottom: 
+    enable: false
+    Options:
+      iconStyle: box
+      boxForm: horizontal
+      position: middleCenter
+      networks: Weibo,Wechat,Douban,QQZone,Twitter,Facebook
```
Comment (ivan-nginx) [inline _config.yml]: Mailto,Twitter,Pinterest,Facebook,GooglePlus,Reddit,Delicious,Tapiture,StumbleUpon,Linkedin,Slashdot,Technorati,Posterous,Tumblr,GoogleBookmarks,Newsvine,Pingfm,Evernote,Friendfeed,Vkontakte,Odnoklassniki,Mailru

File: _config.yml [hunk 9]
```diff
@@ -390,6 +390,26 @@ valine:
...
+  float: 
+    enable: false
+    Options:
+      iconStyle: box
+      boxForm: horizontal
+      position: middleRight
+      networks: Weibo,Wechat,Douban,QQZone,Twitter,Facebook
```
Comment (ivan-nginx) [inline _config.yml]: Mailto,Twitter,Pinterest,Facebook,GooglePlus,Reddit,Delicious,Tapiture,StumbleUpon,Linkedin,Slashdot,Technorati,Posterous,Tumblr,GoogleBookmarks,Newsvine,Pingfm,Evernote,Friendfeed,Vkontakte,Odnoklassniki,Mailru

File: layout/_third-party/share/needsharebutton/floatdom.swig (hunk 6 above)
Comment (LEAFERx) [reply layout/_third-party/share/needsharebutton/floatdom.swig]: It is ok to use macro to identify whether it is float or postbottom?And where should i import the macro in _layout.swig and post.swig?

Comment (ivan-nginx) [reply layout/_third-party/share/needsharebutton/floatdom.swig]: U joking? This is floatdom.**swig**

Comment (LEAFERx) [reply layout/_third-party/share/needsharebutton/floatdom.swig]: I will make a new file called dom.swig like this:
```swig
{% macro render type %}
    <span id="needsharebutton-{{type}}"><i class="fa fa-share-alt" aria-hidden="true"></i></span>
{% endmacro %}
```
(maybe i reply in the wrong place...)

Comment (ivan-nginx) [reply layout/_third-party/share/needsharebutton/floatdom.swig]: Seems good. But why filename `dom.swig`?

**Comment by ivan-nginx** [comment]
Also, take a look on this screen:
image

As u can see, there are [main widget](https://github.com/iissnan/hexo-theme-next/blob/master/source/css/_common/components/post/post-widgets.styl) what include:

1. [Social rating](https://github.com/iissnan/hexo-theme-next/blob/master/layout/_third-party/rating.swig).
2. [VK likes](https://github.com/iissnan/hexo-theme-next/blob/master/layout/_third-party/analytics/vkontakte-api.swig).
3. [FB likes & shares](https://github.com/iissnan/hexo-theme-next/blob/master/layout/_third-party/analytics/facebook-sdk.swig).

**Comment by LEAFERx** [comment]
OK. I'll try.

**Comment by LEAFERx** [comment]
1 lowercase in config
2 put everything into needsharebutton.swig and needsharebutton.styl
3 rewrite the description
4 put the postbottom button into the post-widget
5 test it in Pisces&Gemini and find it works in the right way

**Comment by ivan-nginx** [comment]
Ok, i'll test it now too, wait a moment...

**Comment by ivan-nginx** [comment]
```yml
needmoreshare2:
  enable: true
  postbottom: 
    enable: true
    options:
      iconStyle: box
      boxForm: horizontal
      position: bottomCenter
      networks: Weibo,Wechat,Douban,QQZone,Twitter,Linkedin,Mailto,Twitter,Reddit,Delicious,Tapiture,StumbleUpon,Pinterest,Facebook,GooglePlus,Linkedin,Slashdot,Technorati,Posterous,Tumblr,GoogleBookmarks,Newsvine,Pingfm,Evernote,Friendfeed,Vkontakte,Odnoklassniki,Mailru
  float: 
    enable: true
    options:
      iconStyle: box
      boxForm: horizontal
      position: middleRight
      networks: Weibo,Wechat,Douban,QQZone,Twitter,Linkedin,Mailto,Twitter,Reddit,Delicious,Tapiture,StumbleUpon,Pinterest,Facebook,GooglePlus,Linkedin,Slashdot,Technorati,Posterous,Tumblr,GoogleBookmarks,Newsvine,Pingfm,Evernote,Friendfeed,Vkontakte,Odnoklassniki,Mailru
```

image
image

[It seems](http://test.almostover.ru) some css styles need to improvement.

**Comment by LEAFERx** [comment]
I'm trying but i meet some troubles when making web font.
It may take some time since I don't have much time now.

**Comment by LEAFERx** [comment]
I'm going to make webfont on fontello like what revir does, but i find that Tapiture,Slashdot,Technorati,Googlebookmarks,Nwsvine,Pingfm,Mailru dont have icons on fontello.
Is it acceptale to remove those social networks?

**Comment by ivan-nginx** [comment]
@LEAFERx hi again.

* Tapiture
* Slashdot
* Technorati
* Googlebookmarks
* Nwsvine
* Pingfm
* Mailru

> Is it acceptale to remove those social networks?

I think yes, but we must try to add all of possible icons if this options already exists and created.
Well, i think need a little wait for answer from revir's issue.

Plugin is really good and i think need to modify him with greatest usability and options.

**Comment by LEAFERx** [comment]
I've fixed the icon problem by a hacky way that is put two webfont file together. But it's not the best way because I'm not sure how many icons are repeated in two files. Should I commit this fix?
BTW, I still think that I should put the min file because other libs are minified like pace, three and so on. But if you insist, I'll put the non-min file.

**Comment by ivan-nginx** [comment]
> by a hacky way that is put two webfont file together. But it's not the best way because

Yea, i agree with this.

> BTW, I still think that I should put the min file because other libs are minified like pace, three and so on. But if you insist, I'll put the non-min file.

All of JS and CSS files in NexT are non-minified. This was do for better maintenance, but it for NexT-libraries at least. For *vendors* (lib directory) libs there are min and non-min libs.
I think the better solution is to place both (min and non-min, 4 files (2 js and 2 css)) and use min versions. As will need to edit some of they, we can see code directly and after edit, for example, can to minify and use edited version.

**Comment by LEAFERx** [comment]
How can i embed a font file into a css?

**Comment by ivan-nginx** [comment]
U mean what code (`\e800`) will have font?

**Comment by ivan-nginx** [comment]
@LEAFERx > `add seperate line and make some refactors in css`

No need to do this. I already refactor this styles but do not finish some button's mastering. U need just concentrate on JS and Fonts, ok?
For example, Hound report u about errors. So, need to fix they. And need to do test it on non-minify files. And only after that do minify. After all work finished and before global merge to NexT master, we just switch from non-minify to minified file. That's what i'm talking about u: no need to add minification — minification do hard to maintenance and only good for prodaction; but for prod's anyone may use minifications plugins.

**Comment by LEAFERx** [comment]
I think the js is ok now

**Comment by ivan-nginx** [comment]
«Line is too long» — just place in new line like that:

Before:
```js
width = window.innerWidth ? window.innerWidth : document.documentElement.clientWidth ? document.documentElement.clientWidth : screen.width,
```

After:
```js
width = window.innerWidth ?
        window.innerWidth :
        document.documentElement.clientWidth ?
        document.documentElement.clientWidth :
        screen.width,
```
It's better for vision. Houndy is smart. :smirk:
https://github.com/iissnan/hexo-theme-next/blob/217b4d857f5fa0b9bee00bb16d021f779923b693/.jshintrc#L10

**Comment by LEAFERx** [comment]
lines is cut now

**Comment by ivan-nginx** [comment]
See? Hound is happy now. :smiley:

**Comment by LEAFERx** [comment]
Please test it if you have time

**Comment by ivan-nginx** [comment]
Already done? Sure, i just finishing my test cases with buttons. Take a moment..

**Comment by ivan-nginx** [comment]
Check it. If all alright, need to do some mastering and finally — merge it.

**Comment by ivan-nginx** [comment]
## Bugs
* In Gemini scheme with post button when click — social show under button, ok; when click again (or if click on any other place) — social dissapear, but blinked in the bottom (change default show place).

**Comment by LEAFERx** [comment]
I've remade the fontello font file and i can say that there is no repeat in two font files.
The blink should be the css animation problem.

**Comment by ivan-nginx** [comment]
@LEAFERx alright, dude, my time is to late, need to sleep. I believe in u remastering this cool things and we create best social plugin for NexT. :+1:

**Comment by LEAFERx** [comment]
blink bug is fixed

**Comment by ivan-nginx** [comment]
Hi! Some small bugs found:

* Titles on social icons when mouse move on them — lowercases. Can we do, for example, instead of
  ```html
  <span class="need-share-button_link-box ... title="weibo"></span>
  ```
  Like this:
    ```html
  <span class="need-share-button_link-box ... title="Weibo"></span>
  ```
  ?

**Comment by ivan-nginx** [comment]
> Identifier 'network_lc' is not in camel case.

`network_lc` » `networkLC`

**Comment by ivan-nginx** [comment]
So, need to resolve conflicts, test it and can be merged i think.

P.S. Sorry for conflicts, it's my mistake with branches and new gui soft. See last commits in master branch.

**Comment by LEAFERx** [comment]
Sorry but it was the first time i merge such a complex branch
the git auto-merging deletes some lines in some files without noticing me and i really dont know if i left something
can you please resolve the conflicts?

**Comment by ivan-nginx** [comment]
Yeah, i just don't understand what's the problem with this merge, all was fine just yesterday.

> can you please resolve the conflicts?

image

But i can make another: i can merge your this branch directly in master or dev branch to solve it, but need to do end of the maintenance. Merge ready? All possible bugs and another things fixed? Need to delete all commented lines and make code clean in `source/css/_common/components/third-party/needsharebutton.styl`, for example. So, please, finish it and i'll try to push this merge directly.

image
The program is «GitKraken». It's free and have nice GUI. Recommended for use.

**Comment by LEAFERx** [comment]
I find a strange css bug
the button has different style and hover behavior in four theme scheme
in muse it is black background and white hover background
in mist it is always white background whether hover or not
in pisces&gemini it is white background and black hover background

**Comment by ivan-nginx** [comment]
It's not a bug, that styles reused from main buttons NexT styles and on different schemes — different styles. See live demo with buttons [here](http://test.almostover.ru/button-tag-test/).

Ok, do not edit for now anymore. I just try to resolve this conflicts and merge it into master, afterthat we can resume to fix all bugs, etc.

**Comment by LEAFERx** [comment]
ok

**Comment by ivan-nginx** [comment]
@LEAFERx check please [button-refactor](https://github.com/iissnan/hexo-theme-next/compare/button-refactor) branch. All NeedMoreShare options there? All latest commits, etc.?

**Comment by LEAFERx** [comment]
yes, and I think it's ready to merge

**Comment by ivan-nginx** [comment]
Ok, i'll do it soon.

P.S. Conflict was. i think, because u or i don not **stage** pushes into your commit directly. It was my fault, i was must to create separated pull or branch with my own button refactor changes:
image

No need to create so must difficult commits in next time.

**Comment by ivan-nginx** [comment]
Alright, branch was merged. Any additions/features/fixes to this pull are welcome.

Anyway, god job @LEAFERx! Also greatings thanks for Gitment pull and halp with adaptation of Gitmint! I really appreciate it.

**Comment by ivan-nginx** [comment]
About buttons: well, it's inherit styles from main readmore buttons and buttons tags too. Need just redefine in `_custom.styl` to `.btn` class any style u want — all buttons (and share button too) will show in this styles.

**Comment by ivan-nginx** [comment]
@LEAFERx hi again! How are u? Fine? Ok, there is 2 bugs:

1. Then user click on any share button and share it, buttons panel after that don't close automatically.
2. Facebook share not work:
    image
3. In mobilw style icons not flexed in main window:
    image

**Comment by LEAFERx** [comment]
Sorry that i dont have time to fix it now, my school is begin and a important exam is waiting for me.
No until Nov will i have time i guess.
Sorry again.

**Comment by ivan-nginx** [comment]
Ok, no problem, dude. Just take a time and try to fix it later. No need to hurry somewhere, it's opensource.

**Comment by online** [comment]
@ivan-nginx, как по мне, то нужно добавить условие, чтобы NeedMoreShare2 (боковая кнопка, которая прилипает к левой стороне браузера) не выводился на главной странице, а только на страницах заметок блога. Что скажешь?

**Comment by ivan-nginx** [comment]
@online так в этом и прикол, что он сделал 2 кнопки:

1. Боковая левая выводится везде.
2. Кнопка для постов выводится в постах.

Вообще, заметил, что всё же соц-иконы Бирмана выдают цифру репостов некоторых. Скорее всего интегрирую и их тоже + там также через JS, оказывается, сделано (то бишь ссылки индекс и вес страниц забирать не будут).

**Pull request merged**